Use a single health log table

MrZammler opened the pull request:
##### Summary
<!--
Describe the change in summary section, including rationale and design decisions.
Include "Fixes #nnn" if you are fixing an existing issue.
-->

This PR changes health to use a single `health_log` table, instead of a table for each host.

The PR will update database to version 9 and will move data from all the `health_log_XXXX` tables to the new one. A new `host_id` column is added, which is the host's `host_uuid`.

##### Test Plan

<!--
Provide enough detail so that your reviewer can understand which test cases you
have covered, and recreate them if necessary. If our CI covers sufficient tests, then state which tests cover the change.
-->

After this PR all of health functionality (including cleaning up and cloud connectivity) should remain as before.

##### Additional Information
<!-- This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue. -->

<details> <summary>For users: How does this change affect me?</summary>
  <!--
Describe the PR affects users: 
- Which area of Netdata is affected by the change?
- Can they see the change or is it an under the hood? If they can see it, where?
- How is the user impacted by the change? 
- What are there any benefits of the change? 
-->
</details>

stelfrag commented on database/sqlite/sqlite_health.c:
```diff
@@ -11,30 +11,29 @@
 */
-#define SQL_CREATE_HEALTH_LOG_TABLE(guid) "CREATE TABLE IF NOT EXISTS health_log_%s(hostname text, unique_id int, alarm_id int, alarm_event_id int, config_hash_id blob, updated_by_id int, updates_id int, when_key int, duration int, non_clear_duration int, flags int, exec_run_timestamp int, delay_up_to_timestamp int, name text, chart text, family text, exec text, recipient text, source text, units ...
-int sql_create_health_log_table(RRDHOST *host) {
+#define SQL_CREATE_HEALTH_LOG_TABLE "CREATE TABLE IF NOT EXISTS health_log (hostname text, host_id blob, unique_id int, alarm_id int, alarm_event_id int, config_hash_id blob, updated_by_id int, updates_id int, when_key int, duration int, non_clear_duration int, flags int, exec_run_timestamp int, delay_up_to_timestamp int, name text, chart text, family text, exec text, recipient text, source text, ...
```
We probably don't need the hostname anymore in the table

MrZammler replied:
Useful for debugging, but yes, will remove.

I'll check to see if we can also remove the `source` field. Likely can be stored along with alert hashes and we can join there to get it when we need it.

stelfrag commented on database/sqlite/sqlite_health.c:
```diff
@@ -43,27 +42,21 @@ int sql_create_health_log_table(RRDHOST *host) {
     if (unlikely(rc != SQLITE_OK)) {
-        sql_create_health_log_table(host);
-        rc = sqlite3_prepare_v2(db_meta, command, -1, &res, 0);
+        sql_create_health_log_table(); //might remove
```
Probably create table during startup always and decide migration as needed. These additional checks here can be removed

MrZammler commented:
What has changed with the PR:

* Create a single `health_log` table. Creating happens during db init, so no need to re-try from various places.
* The new table adds a new field `insert_mark_timestamp`. It is a `monotonic_usec` added when there is an insert. We'll see if it does what we want.
* A check during filtering of alert events to be streamed to cloud has been removed, since that check was based on the assumption that the REMOVED events where created during shutdown which is no longer the case.
* Added the `source` field to `alert_hash` which stores the line number and filename of the alert.
* Removed fields `hostname`, `source`, `class`, `component` and `type` from the `health_log` table. Those can be retrieved from the `alert_hash` table. This should reduce the database size for same amount of stored events.

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -45,24 +16,31 @@ void update_filtered(ALARM_ENTRY *ae, uint32_t unique_id, char *uuid_str) {
-#define SQL_SELECT_ALERT_BY_UNIQUE_ID "SELECT hl.unique_id FROM health_log_%s hl, alert_hash ah WHERE hl.unique_id = %u " \
-                            "AND hl.config_hash_id = ah.hash_id " \
+#define SQL_SELECT_ALERT_BY_UNIQUE_ID "SELECT hl.unique_id FROM health_log hl, alert_hash ah WHERE hl.unique_id = %u " \
+                            "AND hl.config_hash_id = ah.hash_id AND host_id = @host_id " \
```
Why are we not using an `INNER JOIN` ? I will write without to test the new query:

```sh
SELECT hl.unique_id FROM health_log hl INNER JOIN alert_hash ah ON(hl.config_hash_id = ah.hash_id) WHERE hl.unique_id = %u  AND host_id = @host_id;
```

stelfrag replied:
It is ok to use the older syntax : from the sqlite documentation https://www.sqlite.org/optoverview.html#joins
```
The ON and USING clauses of an inner join are converted 
into additional terms of the WHERE clause prior to WHERE 
clause analysis described [above in paragraph 2.0]
Thus with SQLite, there is no computational advantage to 
use the newer SQL92 join syntax over the older SQL89 comma-join syntax. 
They both end up accomplishing exactly the same thing on inner joins
```

thiagoftsm replied:
Thank you to bring this additional info.

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -307,19 +274,15 @@ void aclk_push_alert_event(struct aclk_sync_host_config *wc)
-        " where hl.unique_id = aa.alert_unique_id and aa.date_submitted is null " \
-        " order by aa.sequence_id asc limit %d;", wc->uuid_str, wc->uuid_str, limit);
+        " from health_log hl, aclk_alert_%s aa, alert_hash ha " \
+        " where hl.unique_id = aa.alert_unique_id and hl.config_hash_id = ha.hash_id and aa.date_submitted is null and host_id = @host_id " \
```
If we do not have, please, create an index that maps all `date_submitted = NULL`. I am not sure whether sqlte accepts [partial index](https://www.postgresql.org/docs/current/indexes-partial.html), if it does not accept, creates an index for the field.
Probably I do not need to say nothing about `INNER JOIN`. :)

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -486,17 +456,49 @@ void sql_queue_existing_alerts_to_aclk(RRDHOST *host)
+    netdata_rwlock_wrlock(&host->health_log.alarm_log_rwlock);
 
-    netdata_rwlock_rdlock(&host->health_log.alarm_log_rwlock);
+    buffer_sprintf(sql, "delete from aclk_alert_%s; ", uuid_str);
```
I already read papers saying that `delete` is faster than `truncate`, on the other hand some databases when we delete the data is still kept hidden, so few questions:

- Did we benchmark `delete` vs`truncate` on SQLite? (I am ignoring `drop`, because it should be more expensive)
- considering that we are deleting `aclk_alert` content, I understand we are not unifying them, do we plan to unfy in the future?
- Are we going to run `vacuum` on these tables?

stelfrag replied:
There is no truncate but there is a special delete optimization for this case. 
```
When the WHERE clause and RETURNING clause are both omitted from 
a DELETE statement and the table being deleted has no triggers, SQLite 
uses an optimization to erase the entire table content without having to visit 
each row of the table individually. 
```

thiagoftsm replied:
Thank you again! :)
The Lite is really light :laughing:

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -486,17 +456,49 @@ void sql_queue_existing_alerts_to_aclk(RRDHOST *host)
+    buffer_flush(sql);
+    buffer_sprintf(sql, "insert into aclk_alert_%s (alert_unique_id, date_created, filtered_alert_unique_id) " \
+                   "select unique_id alert_unique_id, unixepoch(), unique_id alert_unique_id from health_log " \
+                   "where new_status <> 0 and new_status <> -2 and config_hash_id is not null and updated_by_id = 0 " \
```
do we have indexes for `new_status`, `config_hash_id` and `updated_by_id`? 
If I see the create table in next lines I will update/remove this comment.

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -685,8 +687,8 @@ void aclk_start_alert_streaming(char *node_id, bool resets)
 #define SQL_QUEUE_REMOVE_ALERTS "INSERT INTO aclk_alert_%s (alert_unique_id, date_created, filtered_alert_unique_id) " \
-                                "SELECT unique_id alert_unique_id, UNIXEPOCH(), unique_id alert_unique_id FROM health_log_%s " \
-                                "WHERE new_status = -2 AND updated_by_id = 0 AND unique_id NOT IN " \
+                                "SELECT unique_id alert_unique_id, UNIXEPOCH(), unique_id alert_unique_id FROM health_log " \
```
In previous query we had `unixepoch()` in lowercase, here we have in uppercase, I suggest to use only one pattern for functions.

thiagoftsm commented on database/sqlite/sqlite_aclk_alert.c:
```diff
@@ -685,8 +687,8 @@ void aclk_start_alert_streaming(char *node_id, bool resets)
+                                "WHERE host_id = @host_id AND new_status = -2 AND updated_by_id = 0 AND unique_id NOT IN " \
                                 "(SELECT alert_unique_id FROM aclk_alert_%s) " \
                                 "AND config_hash_id NOT IN (select hash_id from alert_hash where warn is null and crit is null) " \
                                 "ORDER BY unique_id ASC " \
```
You organize `unique_id` as ASC in the external query, but in the inner you did not organize them, probably, ordering both will simplify database's life.

thiagoftsm commented on database/sqlite/sqlite_db_migration.c:
```diff
@@ -214,6 +213,63 @@ static int do_migration_v7_v8(sqlite3 *database, const char *name)
+              "transition_id blob, global_id int);");
+    sqlite3_exec_monitored(database, sql, 0, 0, NULL);
+
+    snprintfz(sql, 2047, "CREATE INDEX IF NOT EXISTS health_log_index ON health_log (unique_id);");
```
We query for more than one value, so we should have more indexes.

stelfrag replied:
Agree need additional

thiagoftsm commented on database/sqlite/sqlite_db_migration.c:
```diff
@@ -214,6 +213,63 @@ static int do_migration_v7_v8(sqlite3 *database, const char *name)
+              "flags int, exec_run_timestamp int, delay_up_to_timestamp int, name text, chart text, family text, exec text, " \
+              "recipient text, units text, info text, exec_code int, new_status real, old_status real, delay int, " \
+              "new_value double, old_value double, last_repeat int, chart_context text, " \
+              "transition_id blob, global_id int);");
```
Only for we do not forget, but as we (@stelfrag , @MrZammler and me) already talked in the past different times, this table is necessary but it not sufficient. I cannot see atomicity in this table and consiering that neither primary key was made, so there are margin to improve. :thinking:

MrZammler replied:
Ok as also discussed with @stelfrag , I'll do an effort to break `health_log` into another table, `health_log_details`. The aim will be to keep basically one row per alert in `health_log`, having events go to the `health_log_details`.

Items such as `chart`, `name`, `chart_context`, `units`, etc will remain in `health_log`, while items like `new_value`, `old_value`, `statuses`, etc to go to `health_log_details`, thus minimizing repeating items. @thiagoftsm how does this sound?

thiagoftsm replied:
Hello @MrZammler ,

I understand what you wrote as a good step in the right direction . And I hope before migration 10 we can have a table for `host`, another for `alerts` and finally a relationship `m <-> n` for them. :)

thiagoftsm commented on database/sqlite/sqlite_functions.c:
```diff
@@ -43,6 +45,14 @@ const char *database_config[] = {
+    "config_hash_id blob, updated_by_id int, updates_id int, when_key int, duration int, non_clear_duration int, "
+    "flags int, exec_run_timestamp int, delay_up_to_timestamp int, name text, chart text, family text, exec text, "
+    "recipient text, units text, info text, exec_code int, new_status real, old_status real, delay int, "
+    "new_value double, old_value double, last_repeat int, chart_context text, transition_id blob, global_id int);",
```
We have a lot of `blob` in our tables, so I understand we are storing binaries here. If it is not the case, the ideal scenario would be to convert for `VARCHAR(36)` or `TEXT`.

stelfrag replied:
blobs in this case store 16 bytes (UUIDs) instead of 36

thiagoftsm replied:
Thank you!

thiagoftsm commented on database/sqlite/sqlite_health.c:
```diff
@@ -689,12 +660,11 @@ void sql_check_removed_alerts_state(char *uuid_str)
    Load from the health log table
 */
-#define SQL_LOAD_HEALTH_LOG(guid) "SELECT hostname, unique_id, alarm_id, alarm_event_id, config_hash_id, updated_by_id, updates_id, when_key, duration, non_clear_duration, flags, exec_run_timestamp, delay_up_to_timestamp, name, chart, family, exec, recipient, source, units, info, exec_code, new_status, old_status, delay, new_value, old_value, last_repeat, class, component, type, chart_context, tr ...
+#define SQL_LOAD_HEALTH_LOG "SELECT hl.unique_id, hl.alarm_id, hl.alarm_event_id, hl.config_hash_id, hl.updated_by_id, hl.updates_id, hl.when_key, hl.duration, hl.non_clear_duration, hl.flags, hl.exec_run_timestamp, hl.delay_up_to_timestamp, hl.name, hl.chart, hl.family, hl.exec, hl.recipient, ah.source, hl.units, hl.info, hl.exec_code, hl.new_status, hl.old_status, hl.delay, hl.new_value, hl.old ...
```
Like one of previous comment, I suggest to reorganize fields per table.

thiagoftsm commented on database/sqlite/sqlite_health.c:
```diff
@@ -1208,7 +1187,7 @@ int sql_health_get_last_executed_event(RRDHOST *host, ALARM_ENTRY *ae, RRDCALC_S
 }
 
-#define SQL_SELECT_HEALTH_LOG(guid) "SELECT hostname, unique_id, alarm_id, alarm_event_id, config_hash_id, updated_by_id, updates_id, when_key, duration, non_clear_duration, flags, exec_run_timestamp, delay_up_to_timestamp, name, chart, family, exec, recipient, source, units, info, exec_code, new_status, old_status, delay, new_value, old_value, last_repeat, class, component, type, chart_context,  ...
+#define SQL_SELECT_HEALTH_LOG "SELECT hl.unique_id, hl.alarm_id, hl.alarm_event_id, hl.config_hash_id, hl.updated_by_id, hl.updates_id, hl.when_key, hl.duration, hl.non_clear_duration, hl.flags, hl.exec_run_timestamp, hl.delay_up_to_timestamp, hl.name, hl.chart, hl.family, hl.exec, hl.recipient, ah.source, hl.units, hl.info, hl.exec_code, hl.new_status, hl.old_status, hl.delay, hl.new_value, hl.o ...
```
if you accept suggestion to reorganize, please, reorganize this too.

thiagoftsm reviewed (approved):
PR is working as expected, LGTM!
